text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+
+- Start Date: 2022-08-12
+- RFC PR: (leave this empty)
+- FOLIO Issue: (leave this empty)
+
+# Cease Creating Kafka Topics For Each Tenant
+
+## Summary
+
+The purpose of this RFC is to lay out the reasoning for creating Kafka topics using a different scheme other than what is currently employed; creating the same topic for each tenant. The recommendation is to create single topic for each application concern. This will reduce the partition count needed for normal operations.
+
+## Motivation
+FOLIO should cease creating Kafka topics for each tenant installed. This has scaling & cost implications.
+
+- **Performance**: 
+Similar to a database not having a theoretical limit on the number of rows/tables, Kafka does not have a limit on partition count. A major difference is that a partition is "heavier" than a row/table. Partitions have to be rebalanced, replicated with corresponding open server files for each partition. Partitions have an ongoing administrative cost while online. Creating numerous partitions like they are database rows/tables will spell disaster for an under-provisioned Kafka installation. This is why cloud providers ensure to have limits on partition counts. Provisioning a Kafka installation to support such a high partition count will be wasteful for FOLIO hosting providers.
+- **Cost**: Kafka's use in FOLIO is still limited, more FOLIO modules will continue to employ event driven techniques that will cause an explosion of partitions. Adding more brokers to service the load compounded by partitions will mean increased costs.
+
+## Out of Scope
+- Enabling tenant data security within a Kafka installation. Tenant data security meaning "How do we prevent a tenant's data to be processed in place of another?", "How can we prevent a tenant module from seeing data from other tenants?"
+- Moving forward with the [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw).
+- Countering module multi-versioning i.e. more than one version of a module is installed in FOLIO with tenants able to target specific versions.
+- FOLIO modules that do not use folio-kafka-wrapper to interact with Kafka. These modules should be migrated to folio-kafka-wrapper to ease partition counts.
+
+## Rationale and Alternatives
```

Comment:
Number of partitions for a topic is not defined within FOLIO code. That configuration is a concern that is shared with a Kafka administrator which will depend on several factors to support high availability and performance. Topic definition is within FOLIO code and can be controlled. FOLIO needing one topic instead of ten topics, will reduce the partition counts regardless of what configuration an administrator may choose. Do you still want me to put something down?

Reply:
> Number of partitions for a topic is not defined within FOLIO code. That configuration is a concern that is shared with a Kafka administrator which will depend on several factors to support high availability and performance. 

Oh, I thought that FOLIO created the topics and set the partition count for the topic. Maybe I've misunderstood how Kafka configuration is done.

> Topic definition is within FOLIO code and can be controlled. FOLIO needing one topic instead of ten topics, will reduce the partition counts regardless of what configuration an administrator may choose. Do you still want me to put something down?

I don't know. In my mind, reducing the partition count per topic was a viable alternative solution that would be far less disruptive. It seems I may have misunderstood the relationship between topics and partition counts and thus am unsure.

Reply:
[This code](https://github.com/folio-org/folio-kafka-wrapper/blob/bf7d3f154f11af3891c4779627c27e75514aa5c3/src/main/java/org/folio/kafka/services/KafkaAdminClientService.java#L88) in folio-kafka-wrapper suggests that FOLIO does define the partition count for a topic when creating it and currently uses 50 for all topics.

Thus, it would require reconfiguration of the topics by the administrator if they wanted to change this.

Have I misinterpreted this code?

Reply:
Yes, that code is available and is used to assign partition counts. Mod-search does this. But not all modules create a topic before using it. If the auto-create feature was disabled, some modules will need to be modified. I believe mod-pub-sub and most data import modules falls into this camp. I have added a section in the doc to cater to this option. Thanks for bringing it up!

Reply:
> not all modules create a topic before using it. If the auto-create feature was disabled, some modules will need to be modified. I believe mod-pub-sub and most data import modules falls into this camp. I have added a section in the doc to cater to this option.

Oh, I did not know that.

It would be good for FOLIO modules to be consistent. Either require topics to be created in advance or always created by the module explicitly. I would prefer auto-create was disabled.

I guess it's out of scope of this work (though could be a factor in the chosen solution).